framework/src/base/NonlinearEigenSystem.C

```diff
@@ -42,14 +43,46 @@ assemble_matrix(EquationSystems & es, const std::string & system_name)
   p->computeJacobian(
       *eigen_system.current_local_solution.get(), *eigen_system.matrix_A, Moose::KT_NONEIGEN);
 
+  Mat petsc_mat_A = static_cast<PetscMatrix<Number> &>(*eigen_system.matrix_A).mat();
+
+  PetscObjectComposeFunction(
+      (PetscObject)petsc_mat_A, "formJacobian", Moose::SlepcSupport::mooseSlepcEigenFormJacobianA);
+  PetscObjectComposeFunction(
+      (PetscObject)petsc_mat_A, "formFunction", Moose::SlepcSupport::mooseSlepcEigenFormFunctionA);
+
+  PetscContainer container;
+  PetscContainerCreate(eigen_system.comm().get(), &container);
+  PetscContainerSetPointer(container, p);
+  PetscObjectCompose((PetscObject)petsc_mat_A, "formJacobianCtx", NULL);
+  PetscObjectCompose((PetscObject)petsc_mat_A, "formJacobianCtx", (PetscObject)container);
+  PetscObjectCompose((PetscObject)petsc_mat_A, "formFunctionCtx", NULL);
+  PetscObjectCompose((PetscObject)petsc_mat_A, "formFunctionCtx", (PetscObject)container);
+
   if (eigen_system.generalized())
   {
     if (eigen_system.matrix_B)
+    {
       p->computeJacobian(
           *eigen_system.current_local_solution.get(), *eigen_system.matrix_B, Moose::KT_EIGEN);
+
+      Mat petsc_mat_B = static_cast<PetscMatrix<Number> &>(*eigen_system.matrix_B).mat();
+
+      PetscObjectComposeFunction((PetscObject)petsc_mat_B,
+                                 "formJacobian",
+                                 Moose::SlepcSupport::mooseSlepcEigenFormJacobianB);
+      PetscObjectComposeFunction((PetscObject)petsc_mat_B,
+                                 "formFunction",
+                                 Moose::SlepcSupport::mooseSlepcEigenFormFunctionB);
+
+      PetscObjectCompose((PetscObject)petsc_mat_B, "formFunctionCtx", NULL);
+      PetscObjectCompose((PetscObject)petsc_mat_B, "formFunctionCtx", (PetscObject)container);
+      PetscObjectCompose((PetscObject)petsc_mat_B, "formJacobianCtx", NULL);
```

Comment:
Well of course you do, `nullptr` is C++11 or above. The reason this is important is that `nullptr` is actually a pointer type whereas, `NULL` is a typedef to 0. In C++11 where you have things like `auto` and initializer lists, having the correct type is important. There are places where NULL can cause issues because the LHS of an expression is expecting to receive the type info from the RHS (note that this never is an issue in C or C++03).